### ISSUE TITLE
Migrating from ARC GNU 2014.12 to 2017.03, differences in final ELF

### ISSUE THREAD

**Issue by claudiotong** [foss-for-synopsys-dwc-arc-processors/toolchain#128]
Hello,

I have a project for ARCv2 EM6 processor and I'm migrating from ARC GNU toolchain 2014.12 to 2017.03, 
After fixing some minor code syntax incompatibilities, I was able to get a successful build, but the produced ELF executable has some differences with the one that I get with the ARC GNU 2014 version.

The attached file readelf_diff.jpg shows the differences between the two ELF files at readelf output (extract).
![readelf_diff](https://user-images.githubusercontent.com/30154509/28181181-f132da08-6807-11e7-90e9-e0fe3fa1a0dd.jpg)

Apart from some minor section size differences (maybe due to improved optimisation), and apart from the appearance of the GNU_STACK segment (that doesn't bother me) the important difference is the allocation of the .fixed section. This section, as the name suggests, is required in my project to be placed at a specific virtual address and physical (load) address: I need VMA=LMA=0x1007fe00 for the .fixed section.

With ARC GNU 2014.12 (file on the left in readelf_diff.jpg), I get the desired result because the .fixed section gets assigned to a separate segment (03) starting at the desired address. The assignment of .fixed to a separate segment also seems to make sense considering the different flags of the sections (.fixed is read-only). 
With ARC GNU 2017.03, instead, the .fixed section is included in the same segment as .rodata and .fill, and to get the same physical address for the .fixed section the segment's FileSize is incremented (from 0x01630 to 0x7aae0) to also include the size of the .fill section, which should take no size on the file being a NOBITS section. This is not desired for my project.

An extract of the linker script that I'm using is in here below:
```bash
MEMORY
{
  CODEROM : ORIGIN = 0x00000000, LENGTH = 512K
  DATAROM : ORIGIN = 0x10000000, LENGTH = 512K
  RAM     : ORIGIN = 0x20000000, LENGTH = 16K
}

ENTRY(_start)

SECTIONS
{
    // sections with VMA in CODEROM, LMA in DATAROM
    .vectors :
    {
        FILL(0);
        . = ALIGN(16);
        *( .vectors);
    } >CODEROM AT>DATAROM
    
    .text :
    {
        FILL(0);
        . = ALIGN(16);
        *(.text);
    } >CODEROM AT>DATAROM

    
    // sections with VMA and LMA in DATAROM
    .rodata :
    {
        FILL(0);
        . = ALIGN(32);
        *(.rodata);
    } >DATAROM

    .fill :
    {
        FILL(0xFFFFFFFF);
        . = ORIGIN(DATAROM) + 0x7fe00;
    } >DATAROM

    .fixed :
    {
        *(.fixed);
    } >DATAROM

    
    // sections in RAM
    .sdata :
    {
        *(.sdata);
    } >RAM

    .data :
    {
        *(.data);
    } >RAM

    .sbss :
    {
        *(.sbss)
    } >RAM

    .bss :
    {
        *(.bss)
        . = ALIGN(4);
    } >RAM

    .heap :
    {
        . = ALIGN(4);
        __start_heap = . ;
        . = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 32))  ;
        __end_heap = . ;
        . = ALIGN(0x8);
    } >RAM

    .stack :
    {
        . = ALIGN(4);
        __stack = . ;
        . = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 2k))  ;
        __stack_top = . ;
    } >RAM

}

```

Is this difference in section-to-segment assignment an intended new behaviour (or bug fix?) or a possible new bug? 
Should I update the linker script (how?) to match the new linker behaviour?

If I modify the following part of the linker script:
`    .fixed :
    {
        *(.fixed);
    } >DATAROM
`
as follows:
`    .fixed (ORIGIN(DATAROM) + 0x7fe00):
    {
        *(.fixed);
    }
`
I get the desired result also with ARC GNU 2017.03, but I'm not sure this is the best way to address the issue.

Thanks a lot in advance, and kind regards.
  Claudio

**Comment by claziss** [comment]
Hi,

A quick question, do u use the same linker script for both toolchains?

Some observations:

- VMA/LMA can be controlled with AT or AT> keyword [1].
- .rodata, .fill, and .fixed are mapped into DATAROM [2]. If you would like to control VMA of .fixed then I would do something like you suggested:

```bash
.fixed ORIGIN(DATAROM) + 0x7fe00:
{
 . = ALIGN (4)
  *(.fixed)
} > DATAROM AT>DATAROM
```

or:

```bash
.fixed ORIGIN(DATAROM) + 0x7fe00:
AT (ORIGIN(DATAROM) + 0x7fe00))
ALIGN (4)
{
  *(.fixed)
} > DATAROM AT>DATAROM
```

Thus you specify the VMA:LMA which are the same. And probably you do not need the fill section.

Additionally, you can mark DATAROM as read only:

```bash
DATAROM (r!w) : ORIGIN = 0x10000000, LENGTH = 512K
```

N.B. The above snips are just guide lines.

References:
[1] [ld manual](https://sourceware.org/binutils/docs/ld/)
[2] [AVR32795](http://www.atmel.com/images/doc32158.pdf)

**Comment by claudiotong** [comment]
Hi Claudiu,

thanks for your reply.

Yes, I'm using the same exact linker script for both toolchain versions, that's why I was surprised by seeing those differences in the ELF.

Based on your suggestions, I removed the .fill section and tried some changes in the .fixed section description in the linker script to see the different results. Here's what I obtained (all trials are done with ARC GNU 2017.03):

1 ) Only VMA specified for .fixed
```c
    .fixed (ORIGIN(DATAROM) + 0x7fe00) :
    {
        *(.fixed);
    }
```
Result: as desired, the .fixed section gets assigned to a separate ELF segment, with the desired VMA/LMA:
```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x002000 0x20000000 0x20000000 0x00000 0x00c28 RW  0x2000
  LOAD           0x002000 0x00000000 0x10000000 0x05520 0x05520 RWE 0x2000
  LOAD           0x007520 0x10005520 0x10005520 0x01600 0x01600 RWE 0x2000
  LOAD           0x100e00 0x1007fe00 0x1007fe00 0x00200 0x00200 R   0x2000
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10

 Section to Segment mapping:
  Segment Sections...
   00     .bss .heap .stack 
   01     .vectors .text
   02     .rodata
   03     .fixed 
   04     
```


2 ) Both VMA and LMA specified for .fixed
```c
    .fixed (ORIGIN(DATAROM) + 0x7fe00) : AT (ORIGIN(DATAROM) + 0x7fe00)
    {
        *(.fixed);
    }
```
Result: same as case 1)


3 ) VMA, VMA region and LMA region specified for .fixed
```c
    .fixed (ORIGIN(DATAROM) + 0x7fe00) :
    {
        *(.fixed);
    } > DATAROM AT>DATAROM
```
Result: almost same results as case 1), but with a very important difference:
the load address (PhysAddr) of the .fixed segment is **not the desired one (0x10006b20 instead of 0x1007fe00)**:
```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x002000 0x20000000 0x20000000 0x00000 0x00c28 RW  0x2000
  LOAD           0x002000 0x00000000 0x10000000 0x05520 0x05520 RWE 0x2000
  LOAD           0x007520 0x10005520 0x10005520 0x01600 0x01600 RWE 0x2000
  LOAD           0x100e00 0x1007fe00 0x10006b20 0x00200 0x00200 R   0x2000
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10

 Section to Segment mapping:
  Segment Sections...
   00     .bss .heap .stack 
   01     .vectors .text
   02     .rodata
   03     .fixed 
   04     
```


4 ) VMA, LMA, VMA region and LMA region specified for .fixed
```c
    .fixed (ORIGIN(DATAROM) + 0x7fe00) : AT (ORIGIN(DATAROM) + 0x7fe00)
    {
        *(.fixed);
    } > DATAROM AT>DATAROM
```
Result: linker error
"section has both a load address and a load region"


5 ) VMA, LMA and VMA region specified for .fixed
```c
    .fixed (ORIGIN(DATAROM) + 0x7fe00) : AT (ORIGIN(DATAROM) + 0x7fe00)
    {
        *(.fixed);
    } > DATAROM
```
Result: same as case 1)

Marking DATAROM as read only as you suggested (`DATAROM (r!w)`) doesn't produce any change in the results.

So, in conclusion, I have some alternatives that give the desired result (case 1, 2, 5), but still I can't understand why the two linker versions (ARC GNU 2017.03 and 2014.12) give different results with the same linker script (the original one, with the .fill section).

Thanks,
  Claudio

**Comment by claziss** [comment]
2014.12 uses a very old binutils version (2.19 if I recall it correctly), 2017.03 uses 2.28. It may explain your observed issue.

In my opinion, the best solution is 5) as you clearly specify where the section is located, besides if .rodata is too large and overlaps .fixed, ld should signalize it through an error.

Best,

**Comment by claudiotong** [comment]
Good point, I agree. I'll go for option 5) then.

I have one last question, not about ELF but still related to the ARC GNU 2014.12 --> 2017.03 transition:

In one assembly file I'm using .extInstruction to specify some extension instructions (GNU Assembler manual, section 9.3.3). One of these instructions is specified as follows:
`.extInstruction instr0, 9, 0x1A, SUFFIX_COND, SYNTAX_2OP`

This was compiling ok without warnings in ARC GNU 2014.12, while in 2017.03 it triggers the following assembler warning message:
`Warning: Suffix SUFFIX_COND ignored`

Also noticed that:
* Other .extInstruction definitions in the same file, that also have SUFFIX_COND but have SYNTAX_3OP instead of _2OP, don't trigger this warning
* The warning disappears if I add the OP1_IMM_IMPLIED modifier:
`.extInstruction instr0, 9, 0x1A, SUFFIX_COND, SYNTAX_2OP|OP1_IMM_IMPLIED`

Does this mean that SUFFIX_COND is only allowed with 3-operand instructions, and hence systematically ignored for 2-operand instructions unless they have an implied immediate destination operand?

Thanks.
  C.

**Comment by claziss** [comment]
Hi there,

I have a wiki on this [subject](https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/wiki/Custom-instructions-what-and-how). As far as I can see, you cannot have conditional execution for 2OP format instruction. Please let me know if I miss something.

**Comment by claziss** [comment]
Actually, it may be possible to use COND for SYNTAX_2OP|OP1_IMM_IMPLIED, which is a SYNTAX_3OP with null destination:
```asm
insn<.cc> limm,c => insn<.cc> 0,limm,c
insn<.cc> limm,limm => insn<.cc> 0,limm,limm
insn<.cc> limm,u6 => insn<.cc> 0,limm,u6
```
Please let me know if those instruction combinations are making sense for you to have them predicated.

Thanks,
C

**Comment by claudiotong** [comment]
Hi,

thanks for the good article you linked. This explains things better.
ARC documentation (ARCv2 ISA Programmer's Reference Manual) seems to confirm that indeed the Conditional suffix <.cc> is allowed only with 3-operand instruction formats (even if it doesn't say why).

What I still don't understand is how to specify a custom instruction that has one source operand and one destination operand, e.g. a byte swap operation that does a <-- swap(b). With reference to your wiki page, should I use the two-operand format **op<.f> b,c** ?
And what if I want to allow a conditional suffix for this instruction? Maybe use the 3-operand format **op<.cc><.f> b,b,c** ?

Thanks

**Comment by claziss** [comment]
It really depends on how the APEX extension is implemented. As you can see in example 3, there are 4 different instructions mapped onto the same major/minor opcode. 
Comming back to your question, I assume your APEX is an instruction with one input and one output, then I would think it is mapped using a clean SYNTAX_2OP format (like abs or swap instructions are):

```asm
.extInstruction instr0, 9,0x1a, SUFFIX_NONE, SYNTAX_2OP
```

If your apex accepts conditional execution field, then you need to go for the full SYNTAX_3OP.

You really need to go back to your hw guys and ask them how that instruction is mapped into your processor.

Best

**Comment by claudiotong** [comment]
I'll do as you suggest. Thanks.

**Comment by claziss** [comment]
I'll close this thread. If you need more info on extension instructions, maybe it is good to make a new one.

Thank you,
Claudiu
